pkg/requestcontext/user_access_control.go

```diff
@@ -0,0 +1,225 @@
+			isAdmin: isAdmin,
+		}
+	} else {
+		uac.authzInfo.isAdmin = isAdmin
+	}
+}
+
+/*
+	UserAcFromContext returns an UserAccessControl struct made available on all http requests
+	(using context.Context values) by authz and authn middlewares.
+
+	If no UserAccessControl is found on context, it will return an empty one.
+
+its methods and attributes can be used in http.Handlers to get user info for that specific request
+(username, groups, if it's an admin, if it can access certain resources).
+*/
+func UserAcFromContext(ctx context.Context) (*UserAccessControl, error) {
+	if uacValue := ctx.Value(GetContextKey()); uacValue != nil {
+		uac, ok := uacValue.(UserAccessControl)
+		if !ok {
+			return nil, errors.ErrBadType
+		}
+
+		return &uac, nil
+	}
+
+	return NewUserAccessControl(), nil
+}
+
+// returns whether or not the user/anonymous who made the request has read permission on 'repository'.
+func (uac *UserAccessControl) CanReadRepo(repository string) bool {
+	if uac.IsAdmin() {
+		return true
+	}
+
+	if uac.authzInfo == nil || uac.authzInfo.readGlobPatterns == nil {
+		return true
+	}
+
+	return uac.matchesRepo(uac.authzInfo.readGlobPatterns, repository)
+}
+
+func (uac *UserAccessControl) SetReadGlobPatterns(patterns map[string]bool) {
+	if uac.authzInfo == nil {
+		uac.authzInfo = &UserAuthzInfo{
+			readGlobPatterns: patterns,
+		}
+	} else {
+		uac.authzInfo.readGlobPatterns = patterns
+	}
+}
+
+func (uac *UserAccessControl) SetDMCGlobPatterns(patterns map[string]bool) {
+	if uac.authzInfo == nil {
+		uac.authzInfo = &UserAuthzInfo{
+			dmcGlobPatterns: patterns,
+		}
+	} else {
+		uac.authzInfo.dmcGlobPatterns = patterns
+	}
+}
+
+/*
+returns whether or not the user/anonymous who made the request
+has detectManifestCollision permission on 'repository'.
+*/
+func (uac *UserAccessControl) CanDetectManifestCollision(repository string) bool {
+	if uac.authzInfo == nil || uac.authzInfo.dmcGlobPatterns == nil {
+		return false
+	}
+
+	return uac.matchesRepo(uac.authzInfo.dmcGlobPatterns, repository)
+}
+
+/*
+returns whether or not 'repository' can be found in the list of patterns
+on which the user who made the request has read permission on.
+*/
+func (uac *UserAccessControl) matchesRepo(globPatterns map[string]bool, repository string) bool {
+	var longestMatchedPattern string
+
+	// because of the longest path matching rule, we need to check all patterns from config
+	for pattern := range globPatterns {
+		matched, err := glob.Match(pattern, repository)
+		if err == nil {
+			if matched && len(pattern) > len(longestMatchedPattern) {
+				longestMatchedPattern = pattern
+			}
+		}
+	}
+
+	allowed := globPatterns[longestMatchedPattern]
+
+	return allowed
+}
+
+func (uac *UserAccessControl) SaveOnRequest(request *http.Request) {
+	uacContext := context.WithValue(request.Context(), GetContextKey(), *uac)
+
+	*request = *request.WithContext(uacContext)
+}
+
+func (uac *UserAccessControl) SaveOnContext(ctx context.Context) context.Context {
```

Comment:
it takes a context as argument, and on top of it it saves a new value on it. in our case userAc struct.

in case we have other values on the arg context, they will not be lost.